src/Workspaces/Core/Portable/CodeActions/CodeAction.cs

```diff
@@ -249,6 +249,17 @@ protected virtual Task<Document> GetChangedDocumentAsync(CancellationToken cance
             return document;
         }
 
+        /// <summary>
+        /// Called by the CodeActions on the UI thread to determine if the CodeAction is still 
+        /// applicable and should be presented to the user.  CodeActions can override this if they 
+        /// need to do any final checking that must be performed on the UI thread (for example
+        /// accessing and querying the Visual Studio DTE).
+        /// </summary>
+        protected internal virtual bool IsApplicable(Workspace workspace)
```

Comment:
Well, '2' at least gives us the ability to do so.  But it does so in a way without pumping.  I don't even know how we could consider pumping.  I mean we could easily get IDE reentrancy with all sorts of editor features *including smart tags* trying to do work.  We will have some severe crazyness if we do that.

If we do want to provide UI access, i think providing a single known point where it is ok would be the preferred route.